src/dac.rs

```diff
@@ -0,0 +1,233 @@
+
+use core::fmt;
+
+use crate::{
+    gpio::{
+        gpioa::{PA4, PA5},
+        Analog,
+    },
+    pac,
+    rcc::APB1,
+};
+
+/// Trait representing a single-channel digital-to-analog converter (DAC).
+pub trait SingleChannelDac<Word> {
+    /// Error type returned by DAC methods
+    type Error;
+
+    /// Output a constant signal, given a bit word.
+    fn try_set_value(&mut self, value: Word) -> Result<(), Self::Error>;
+}
+
+/// This is an abstraction to ensure that the DAC output pin is configured
+/// as an analog output.
+pub trait Pins {}
+impl Pins for PA4<Analog> {}
+impl Pins for PA5<Analog> {}
+
+#[derive(Clone, Copy, Debug)]
+/// Select the channel
+pub enum Channel {
+    /// Channel 1
+    One,
+    /// Channel 2
+    Two,
+}
+
+#[derive(Clone, Copy, Debug)]
+/// Three options are available to set DAC precision.
+pub enum DacBits {
+    /// Eight bit precision, right-aligned.
+    EightR,
+    /// 12-bit precision, left-aligned.
+    TwelveL,
+    /// 12-bit precision, right-aligned.
+    TwelveR,
+}
+
+#[derive(Clone, Copy)]
+/// Select a trigger, used by some features.
+pub enum Trigger {
+    /// Timer 6
+    Tim6,
+    /// Timers 3 or 8
+    Tim3_8,
+    /// Timer 7
+    Tim7,
+    /// Timer 15
+    Tim15,
+    /// Timger 2
+    Tim2,
+    /// Tim4 is Not available on DAC 2.
+    Tim4,
+    /// Eg, for interrupts
+    Exti9,
+    /// A software trigger
+    Swtrig,
+}
+
+impl Trigger {
+    pub fn bits(&self) -> u8 {
+        match self {
+            Self::Tim6 => 0b000,
+            Self::Tim3_8 => 0b001,
+            Self::Tim7 => 0b010,
+            Self::Tim15 => 0b011,
+            Self::Tim2 => 0b100,
+            Self::Tim4 => 0b101,
+            Self::Exti9 => 0b110,
+            Self::Swtrig => 0b111,
+        }
+    }
+}
+
+pub struct Dac {
+    regs: pac::DAC,
+    channel: Channel,
+    bits: DacBits,
+    vref: f32,
+}
+
+impl Dac {
+    /// Create a new DAC instance
+    pub fn new<P: Pins>(
+        regs: pac::DAC,
+        _pins: P,
+        channel: Channel,
+        bits: DacBits,
+        vref: f32,
+    ) -> Self {
+        Self {
+            regs,
+            channel,
+            bits,
+            vref,
+        }
+    }
+
+    /// Enable the DAC.
+    pub fn enable(&mut self, apb1: &mut APB1) {
+        apb1.enr().modify(|_, w| w.dac1en().enabled());
+        match self.channel {
+            Channel::One => self.regs.cr.modify(|_, w| w.en1().enabled()),
+            Channel::Two => self.regs.cr.modify(|_, w| w.en2().enabled()),
+        }
+    }
+
+    /// Disable the DAC
+    pub fn disable(&mut self, apb1: &mut APB1) {
+        match self.channel {
+            Channel::One => {
+                self.regs.cr.modify(|_, w| w.en1().disabled());
+                apb1.enr().modify(|_, w| w.dac1en().disabled());
+            }
+            Channel::Two => {
+                self.regs.cr.modify(|_, w| w.en2().disabled());
+                apb1.enr().modify(|_, w| w.dac2en().disabled());
```

Comment:
This still disables DAC2 although we don't support it anymore. Like with `enable`, this should also changed to `dac1en`